doc/modules/ROOT/pages/python-runtime/knowledge-graph-embeddings.adoc

```diff
@@ -0,0 +1,296 @@
+
+image::python-runtime/transe-formula.svg[width=280]
+
+
+=== `DistMult`
+
+`DistMult` is a semantic matching interaction model.
+This method assumes that the relation between head and tail entities is a dot-product multiplication in the embedding space.
+This method works for modelling symmetric relations.
+
+The formula to compute the score of a triple `(head, relation, tail)` using the embeddings `h`, `r`, and `t`:
+
+image::python-runtime/distmult-formula.svg[width=400]
+
+
+== Considerations
+To effectively train KGE models, several considerations need to be taken into account, including the choice of loss function, sampling methods, and optimization strategies.
+
+
+=== Stochastic Local Closed World Assumption (sLCWA)
+
+Observed triplets in the knowledge graph are considered true.
+The unobserved triplets can be treated differently based on the assumption made.
+
+* Open World Assumption (OWA) assumes that all unobserved facts are unknown.
+
+* Closed World Assumption (CWA) assumes that all unobserved facts are false.
+
+* Local Closed World Assumption (LCWA) assumes that all observed facts are true.
+All corrupted triplets, which are generated by replacing the head or tail entity of a positive triplet, are false.
+
+* Stochastic Local Closed World Assumption (sLCWA) assumes that all observed facts are true.
+Some corrupted triplets are false and some are true.
+The number of corrupted triplets for each true triplet is set by the `negative_sampling_size` parameter.
+
+Knowledge graph embedding models are trained under the sLCWA assumption in current implementation.
+
+
+=== Loss function
+The loss function is crucial for guiding the training process of KGE models.
+It determines how the difference between predicted and actual values is calculated and minimized.
+There are several loss functions that can be used for training KGE models, see below for more details.
+
+
+==== Margin ranking loss
+
+Margin ranking loss is a pairwise loss function that compares the scores based on the difference between the scores of a positive triple and a negative triple.
+When negative sampling size is more than one, the loss is computed for a positive triple and each of its negative triples, and the average loss is computed.
+
+image::python-runtime/mrl.svg[width=300]
+image::python-runtime/delta-value.svg[width=400]
+
+
+==== Negative Sampling Self-Adversarial Loss
+
+Negative Sampling Self-Adversarial Lossfootnote:[Sun, Zhiqing, et al. "Rotate: Knowledge graph embedding by relational rotation in complex space." arXiv preprint arXiv:1902.10197 (2019).] is a setwise loss function that compares the scores based on the difference between the scores of a positive triple and a set of negative triples.
+`loss_function_kwargs` can be used to set the `adversarial_temperature` and `margin` parameters.
+
+
+=== Optimizer
+
+Any pytorch optimizer can be used for training the model.
```

Comment:
I'm worried about this. It really depends on the PyTorch version. I think it's quite dangerous to expose this sort of thing

Reply:
Better, but I think we should probably explicitly list exactly the optimizers we support. Similar for schedulers of course. And what happens if these optimizers changes name between PyTorch versions...

In general I'm not super fond of us exposing the implementation details like this. But it's something Gian wants I guess

Reply:
I think we have to do that, or state all names of the optimizers/schedulers, and exactly what parameters they can take. Let's think about it